### ISSUE TITLE
TestClient Get modify url

### ISSUE THREAD

**Issue by abhubcbc** [Kludex/starlette#1060]
I ran into what seems to be a bug in starlette's get function in TestClient.

I get an unexpected behavior when running the test beneath.
```import pytest
from fastapi import FastAPI
from pydantic import constr
from starlette.testclient import TestClient

r_id = r"^\w+$"
app = FastAPI()
constranedString = constr(regex=r_id)


@app.get("/route/{x}")
async def route(x: constrainedString):
    return x


@pytest.mark.parametrize(argnames="val", argvalues=["0;", ";", "0;0"])
def test_call(val):
    client = TestClient(app)
    resp = client.get(f"/route/{val}")

    assert response.status_code == 422
```

When calling get on the test client with val as path parameter
```
Entering test with:
val="0;"
Calling:
client.get("/route/0;")
```
 I expect that my check will return status code 422, but starlette TestClient uses urllib which modifies the input Before it reaches my check, giving it an unwanted behavior.
```
Entering route(x):
x = "0"
expected:
x = "0;"
```

In production it behaves as expected, and returns status code 422 if you use that same value.
Is it really up to the TestClient to make sure that we do not have characters that are either url reserved or url unsafe etc.?

### Environment

- OS: Linux
- Python version: 3.8.0
- Starlette version: 0.13.6

**EDIT** - Add plain starlette example and explanation

It's been a while since I created this issue, but I looked at it a little bit and I will try to explain the issue better.

In _starlette.testclient.TestClient_ the function _request_ uses _urljoin_ to join _test_client.base_url_ and _url_.

```url = urljoin(self.base_url, url)```

The function _urllib.parse.urljoin_ does more than just joining the _url_ and _base_ (look at _urllib.parse.urlparse_ for example), in this case it has an unwanted side effect and strip trailing _;_.

```
from unittest.mock import patch, MagicMock

import pytest
from starlette.testclient import TestClient


@pytest.mark.parametrize(argnames="url", argvalues=["0;", ";", "0;0", ";0", ";;", "0;;", ""])
def test_diff(url):
    base_url = "http://testserver"
    expected_url = "/".join((base_url, url)) if url else base_url
    test_client = TestClient(app=MagicMock())
    with patch("requests.sessions.Session.request") as session_mock:
        test_client.get(url)
        assert expected_url == session_mock.call_args.args[1]
```

In the test example above, sending in "0;" as _url_ to _test_client.get(url)_ then the _session_mock.request_ function will receive "http://testserver/0" instead of "http://testserver/0;" which is the behavior when using the actual _requests.sessions.Session.request_ function.

TL;DR; When using _starlette.testclient.TestClient_, it does not behave the same way as it's parent _requests.Session_ when calling the function _request_.

**Comment by Kludex** [comment]
PR is welcome.

**Comment by lovelydinosaur** [comment]
This description isn't sufficiently narrowed down for us to treat it as an issue.

We'd need an example that's just plain starlette, rather than FastAPI, and a clearer explanation of why you were expecting a particular behaviour.

**Comment by abhubcbc** [comment]
@tomchristie I have updated the issue with more information and a plain starlette example.

@Kludex I am not sure if I can submit a PR, since I do not know all cases where the TestClient should be used. I would assume that the easiest fix would be to remove the default value for _base_url_, but that would probably break a lot of existing code. So the other way would be to not use _urljoin_ in _starlette.testclient.TestClient.request_ and do it manually or with another lib function.
